code refactor

papaj-na-wrotkach opened the pull request:
Code refactor made by me

Commit:
Tokens JWT

Commit:
Tokens JWT

Xertes0 committed:
Fix CMake not configuring

Project still does not compile.

Commit:
working without JWT

Commit:
Buggy code

Commit:
api works but tokens dont

Xertes0 committed:
Add missing nlohmann_json dependency

Xertes0 committed:
Fix jwt dependency typo

cpp-jwt/1.4 -> jwt-cpp/0.6.0

Xertes0 committed:
Fix runtime error

Comment out unused oatpp jwt component.

Commit:
better code

Commit:
still working on that

Commit:
still working on that

Xertes0 committed:
Add missing sources to CMake `server` target

Target still does not compile.

Commit:
almost ready

Commit:
workinggit add .

Commit:
dopracowywanie kodu

Commit:
JWT token working properly

Commit:
tag fix

Commit:
JWT signature from user dto

Commit:
token admin works

Commit:
parkingSpace endpoint

Commit:
JWT auth

Xertes0 committed:
Fix compile errors

Commit:
endpoint parking space

Commit:
working endpoint parking space

Commit:
fix forgein_key

Commit:
fix signUp

Commit:
ignoring local dir

Commit:
ignoring local dir

papaj-na-wrotkach committed:
Delete unnecessary files

Commit:
information if user is admin

Commit:
Boolean

Commit:
leter

Commit:
authendpoints

Commit:
offer create works

Xertes0 committed:
Look for config file in platform specific paths

Looks for the following directories in order:
 - %APPDATA%/parking_space_sharing
 - $XDG_CONFIG_HOME/parking_space_sharing
 - $HOME/.config/parking_space_sharing

Throws std::runtime_error if none of these environment variables were defined.

Xertes0 committed:
MSVC ignore 4996 - getenv unsafe warning

Commit:
samples

Commit:
fix diffrent names with db

Commit:
fix Status404

Commit:
fix code404

Commit:
fix invalid names db

Commit:
fix res->has...

Commit:
change dto

Commit:
Merge branch 'dev/endpoints' into dev/JWT

Commit:
merge compile

Commit:
described dto

Commit:
reserve for admin endpoinds

Commit:
new auth works (hiden)

Commit:
new poles in dto

Commit:
merge fix

Commit:
Documented auth

Commit:
Delete endpoints works

Commit:
create offer works

papaj-na-wrotkach committed:
don't use lower_case

papaj-na-wrotkach committed:
use #pragma once, refactor everything :), use Boost

papaj-na-wrotkach committed:
make username and email unique

papaj-na-wrotkach committed:
remove Boost from conan.cmake (acquire it from system package manager)

papaj-na-wrotkach committed:
Merge branch 'dev/JWT' into dev/swagger-auth-demo

# Conflicts:
#	server/src/controller/offer_controller.hh
#	server/src/controller/parkingSpace_controller.hh
#	server/src/controller/user_controller.hh
#	server/src/database/pss_db.hh
#	server/src/service/offer_service.cc
#	server/src/service/offer_service.hh
#	server/src/service/parkingSpace_service.cc
#	server/src/service/parkingSpace_service.hh
#	server/src/service/user_service.cc
#	server/src/service/user_service.hh

papaj-na-wrotkach committed:
WiP

papaj-na-wrotkach committed:
Basic CRUD done for Place entity

papaj-na-wrotkach committed:
Make use of fts5, create search endpoint, refactor, change std to 23, multi-file migrations

papaj-na-wrotkach committed:
make "id" an exception to the short name of parameters and variables in .clang-tidy

papaj-na-wrotkach committed:
refactor user entity and related app components

papaj-na-wrotkach committed:
refactor authorization, small fixes and improvements

papaj-na-wrotkach committed:
correct error messages

papaj-na-wrotkach committed:
refactor code related to offer entity

Xertes0 commented on .github/workflows/ci.yml:
Why was CI removed?

papaj-na-wrotkach replied:
Huh? I didn't notice. Maybe it was removed in merge from dev/JWT?

papaj-na-wrotkach replied:
Anyways, I'll restore it

papaj-na-wrotkach replied:
1edec1116772ea429c4c55945d5d135e8e1dfa1f deleted it

Xertes0 commented on .gitignore:
```diff
@@ -31,4 +31,11 @@
+.cache/*
+
+run.sh
+CMakeCache.txt
```
`.github` should not be ignored and as far as I'm concerned there should be not reason for `CMakeCache.txt` to be here.

papaj-na-wrotkach replied:
Should we exclude `run.sh`? What is this anyway?

papaj-na-wrotkach replied:
As for `CMakeCache.txt`. It appears to contain directory with username of one of our collaborators so I think we should exclude it.

Xertes0 replied:
But why does it even exist? To me it looks like someone configured CMake in project's root directory instead of `/build` and didn't remove this file thinking it was relevant.

papaj-na-wrotkach replied:
Okay, I shall delete it and remove it from the `.gitignore`

Xertes0 commented on server/src/auth/AuthHandler.cc:
```diff
@@ -0,0 +1,15 @@
+
+namespace server::auth {
+
+    AuthHandler::AuthHandler(std::shared_ptr<JWT> const &jwt)
```
I think we should pass the shared_ptr by value and then std::move it.

Xertes0 commented on server/src/auth/JWT.cc:
```diff
@@ -0,0 +1,47 @@
+                .set_subject(std::to_string(payload->userId))
+                .set_issued_at(std::chrono::system_clock::now())
+                .set_not_before(std::chrono::system_clock::now())
+                .set_expires_at(std::chrono::system_clock::now() + std::chrono::days{30}) //TODO: move to config
```
It might be a good idea to make `std::chrono::system_clock::now()` a variable, but it's up to you.

Xertes0 commented on CMakeLists.txt:
```diff
@@ -29,6 +30,11 @@ find_package(Threads REQUIRED)
+
+find_package(OpenSSL  3.0.8  REQUIRED)
+find_package(jwt-cpp  0.6.0  REQUIRED)
+find_package(Boost    1.81.0 REQUIRED COMPONENTS json)
```
We don't need Boost's 1.81 features, I think we should change the required minimum version to 1.74 (that's what stable debian ships) to be more portable.

Xertes0 commented on server/src/auth/JWT.hh:
```diff
@@ -0,0 +1,42 @@
+#include <jwt-cpp/traits/boost-json/defaults.h>
+#include <oatpp/web/server/handler/AuthorizationHandler.hpp>
+
+using Traits = jwt::traits::boost_json;
```
Maybe we should move it into the namespace and rename to `JSONTraits`. Just `Traits` could be ambiguous.

papaj-na-wrotkach replied:
I'll change the name to `JsonTraits`

Xertes0 commented on server/src/ErrorHandler.hh:
```diff
@@ -0,0 +1,31 @@
+
+#include "dto/StatusDto.hh"
+
+using ResponseFactory = oatpp::web::protocol::http::outgoing::ResponseFactory;
```
This should be in the namespace.

papaj-na-wrotkach replied:
The `using` clause?

Xertes0 replied:
Yes, otherwise this will appear in the global namespace. It would be okay if it wasn't a header file.

Xertes0 commented on server/src/auth/JWT.cc:
```diff
@@ -0,0 +1,47 @@
+
+namespace server::auth {
+
+    JWT::JWT(
+            oatpp::String const &secret,
+            oatpp::String const &issuer
+    ) : secret_(secret), issuer_(issuer),
```
We are copying the string anyway so we should pass by value.

papaj-na-wrotkach replied:
Why not pass by value and `std::move` it?

Xertes0 replied:
> Why not pass by value and `std::move` it?

That is what I meant.

papaj-na-wrotkach replied:
Okay, so I shall do

papaj-na-wrotkach replied:
I have runtime errors when I change it like this:
```c++
JWT::JWT(String secret, String issuer)
  : secret_(std::move(secret))
  , issuer_(std::move(issuer))
```
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  [oatpp::data::mapping::type::String::operator std::string() const]: Error. Null pointer.
```

Xertes0 replied:
lmao

Xertes0 replied:
I had to change from `String` to `std::string`. Already pushed.

Xertes0 commented on server/src/auth/JWT.hh:
```diff
@@ -0,0 +1,42 @@
+    class JWT {
+
+    public:
+        struct Payload : public oatpp::web::server::handler::AuthorizationObject {
+            oatpp::UInt64 userId;
+            oatpp::Int32 role;
+        };
```
Is there a reason why this structure is inside a class? I think it makes things less readable.

papaj-na-wrotkach replied:
I think it is directly associated with the class so why not...
Also this is a modification of file from example Oat++ project.

Xertes0 replied:
We are referencing this structure outside of `JWT.hh` (all controllers and AuthService), so moving it outside of `JWT` class could be better. As I said, I think it'd be more readable.

papaj-na-wrotkach replied:
moved

papaj-na-wrotkach replied:
858137ecccae0545cbae1b22dc12a94ae245fc50

Xertes0 commented on server/src/component/SwaggerComponent.hh:
```diff
@@ -0,0 +1,33 @@
+                            .setTitle("Parking space sharing")
+                            .setDescription("Project created by trainees of Bakcyl Programowania in 2022/2023.")
+                            .setVersion("0.1.0")
+                            .addServer("http://localhost:8000", "server on localhost")
+                            .addServer("http://0.0.0.0:8000", "server on local ipv4")
```
0.0.0.0 will already open on localhost.

Xertes0 replied:
Also 0.0.0.0 listens on all interfaces so it isn't 'local'.

papaj-na-wrotkach replied:
It should listen on all interfaces by default. That was my intention. I added a server here bc some people will get CORS error when using Swagger UI on `0.0.0.0` with `server on localhost` sever chosen

Xertes0 replied:
Okay, but we should change it to "server on all IPv4 interfaces", or something like that, because it's not just "local ipv4". Same with IPv6 equivalent.

papaj-na-wrotkach replied:
> [IANA](https://en.wikipedia.org/wiki/Internet_Assigned_Numbers_Authority), who allocate IP addresses globally, have allocated the single IP address 0.0.0.0[[1]](https://en.wikipedia.org/wiki/0.0.0.0#cite_note-1) to [RFC](https://en.wikipedia.org/wiki/RFC_(identifier)) [1122](https://datatracker.ietf.org/doc/html/rfc1122) section 3.2.1.3.
> 
> It is named there as "This host on this network"

Quoted from [Wikipedia](https://en.wikipedia.org/wiki/0.0.0.0#Official_Standard_Meaning_and_Use)

papaj-na-wrotkach replied:
I guess I'll change it to "This host on this network"

Xertes0 replied:
I've read up on this on `ip(7)` man page, and `0.0.0.0` is said to "bind to all *local* interfaces" so maybe we should leave it as it is with "server on local ipv4". Sorry for misleading.

Xertes0 commented on server/src/auth/AuthHandler.hh:
```diff
@@ -0,0 +1,24 @@
+    private:
+        std::shared_ptr<JWT> jwt_;
+    public:
+        explicit AuthHandler(
```
We use [[nodiscard]] everywhere so might as well use it here.

Xertes0 commented on server/src/controller/OfferController.hh:
```diff
@@ -0,0 +1,201 @@
+#include "service/OfferService.hh"
+#include "service/PlaceService.hh"
+
+using HttpError = oatpp::web::protocol::http::HttpError;
```
Here as well, we should move it to the namespace.

Xertes0 commented on server/src/controller/OfferController.hh:
```diff
@@ -0,0 +1,201 @@
+        ENDPOINT("GET", "offer", search,
+                 QUERY(String, query, "query", std::string{}),
+                 QUERY(UInt64, limit, "limit", 20U),
+                 QUERY(UInt64, offset, "offset", uint64_t{0})) {
```
Inconsistencies with 20U and uint64_t.

Xertes0 commented on server/src/controller/PlaceController.hh:
```diff
@@ -0,0 +1,201 @@
+#include "dto/OfferDto.hh"
+#include "service/PlaceService.hh"
+
+using HttpError = oatpp::web::protocol::http::HttpError;
```
Here as well, move it into the namespace.

Xertes0 commented on server/src/controller/PlaceController.hh:
```diff
@@ -0,0 +1,201 @@
+        [[nodiscard]]
+        explicit
+        PlaceController(
+                std::shared_ptr<ObjectMapper> const &objectMapper,
+                std::shared_ptr<auth::JWT> const &jwtObject
```
shared_ptr should be passed by value

papaj-na-wrotkach replied:
> Clang-Tidy: The parameter 'jwtObject' is copied for each invocation but only used as a const reference; consider making it a const reference

Xertes0 replied:
`jwtObject` is used to make AuthHandler which also should take it by value. If you change it in those 2 places, it should stop complaining.

Xertes0 replied:
By second place I mean `auth::AuthHandler`'s constructor.

Xertes0 commented on server/src/controller/PlaceController.hh:
```diff
@@ -0,0 +1,201 @@
+
+        ENDPOINT("GET", "place", search,
+                 QUERY(String, query, "query", std::string{}),
+                 QUERY(UInt64, limit, "limit", 20U),
+                 QUERY(UInt64, offset, "offset", uint64_t{0})) {
```
Same as OfferController.hh, use literals or uint64_t.

Xertes0 commented on server/src/controller/PlaceController.hh:
```diff
@@ -0,0 +1,201 @@
+                        "Cannot create or modify place of another user as a regular user"
+                )
+            } catch (HttpError &error) {
+                if (error.getInfo().status != Status::CODE_404) { throw; }
```
What is the reason for that if.

papaj-na-wrotkach replied:
Because it uses `PlaceService::getOne` method which can throw 404 if place wasn't found. We accept 404 because if entity doesn't exist then we can create it as it's a `PUT` method endpoint.

papaj-na-wrotkach replied:
It also omits the later checks, which slightly improves performance (not really XD).

papaj-na-wrotkach replied:
I think I should even move calling this method before accessing `authObject->userId` (don't need to access the property `ownerId` if the entity doesn't exist)

Xertes0 replied:
`service_->getOne()` throws 404 the later check `(dto->ownerId == nullptr || dto->ownerId == authObject->userId)` is not asserted. Doesn't this mean that we can create a place for someone else?

papaj-na-wrotkach replied:
Oh, that's my mistake, thanks for pointing it out

papaj-na-wrotkach replied:
resolving as there's a new one on recent changes

Xertes0 commented on server/src/controller/UserController.hh:
```diff
@@ -0,0 +1,200 @@
+
+        ENDPOINT("GET", "user", search,
+                 QUERY(String, query, "query", std::string{}),
+                 QUERY(UInt64, limit, "limit", 20U),
+                 QUERY(UInt64, offset, "offset", uint64_t{0})) {
```
Same as in previous controllers. Use literals or uint64_t.

Xertes0 commented on server/src/database/MainDatabase.hh:
```diff
@@ -0,0 +1,131 @@
+    public:
+        [[nodiscard]]
+        explicit
+        MainDatabase(std::shared_ptr<oatpp::orm::Executor> const &executor)
```
The constructor should be moved to .cc file.

Xertes0 replied:
Also shared_ptr should be passed by value not reference.

papaj-na-wrotkach replied:
We don't have a `.cc` file for this class. Should I create it?

Xertes0 replied:
Yeah

Xertes0 commented on server/src/database/MainDatabase.hh:
```diff
@@ -0,0 +1,131 @@
+        MainDatabase(std::shared_ptr<oatpp::orm::Executor> const &executor)
+                : oatpp::orm::DbClient{executor} {
+            oatpp::orm::SchemaMigration migrator{executor};
+            auto migrations =
+                    std::filesystem::directory_iterator{MAIN_DATABASE_MIGRATIONS_PATH}
+                    | std::views::all | std::views::filter([](auto const &entry) {
+                        return entry.is_regular_file() && entry.path().extension() == ".sql";
+                    }) | std::views::transform([](auto const &entry) {
+                        static auto i = 1L;
+                        return std::make_pair(i++, entry);
+                    });
+            for (auto const &[index, migration]: migrations) {
+                migrator.addFile(index, migration.path().string());
+            }
```
Since we are globing the entire directory, it'd be nice to see a log with the name of the file which will be used for migration.

Xertes0 replied:
I doubt that `std::filesystem::directory_iterator` ensures alphabetical order of the files. What would happen if it'd try to apply the migrations in incorrect order? (e.g 002_sample_data.sql then 001_init.sql)

Xertes0 replied:
Also this could be simple for loop. Are we bumping the c++ standard just for this?

Xertes0 replied:
> I doubt that `std::filesystem::directory_iterator` ensures alphabetical order of the files.

[No, it does not.](https://eel.is/c++draft/filesystems#fs.class.directory.iterator.general-8)

papaj-na-wrotkach replied:
> Also this could be simple for loop. Are we bumping the c++ standard just for this?

Boomer detected, opinion rejected

papaj-na-wrotkach replied:
> > I doubt that `std::filesystem::directory_iterator` ensures alphabetical order of the files.
> 
> [No, it does not.](https://eel.is/c++draft/filesystems#fs.class.directory.iterator.general-8)

What should we do  to ensure the order?

papaj-na-wrotkach replied:
> Also this could be simple for loop. Are we bumping the c++ standard just for this?

Not because of that - you can use all of this in C++20 but I want to use the latest standard if it's not a problem.

Xertes0 replied:
> > Also this could be simple for loop. Are we bumping the c++ standard just for this?
> 
> Not because of that - you can use all of this in C++20 but I want to use the latest standard if it's not a problem.

If we don't use anything from the latest standard, then it is better to stick with c++20 so we can support more compiler versions.

Xertes0 replied:
> > > I doubt that `std::filesystem::directory_iterator` ensures alphabetical order of the files.
> > 
> > 
> > [No, it does not.](https://eel.is/c++draft/filesystems#fs.class.directory.iterator.general-8)
> 
> What should we do to ensure the order?

It think that specifying the migrations by hand would be best.
So just something like this:
```c++
migrator.addFile(1, MAIN_DATABASE_MIGRATIONS_PATH "/001_init.sql");
migrator.addFile(2, MAIN_DATABASE_MIGRATIONS_PATH "/002_sample_data.sql");
```

papaj-na-wrotkach replied:
I disagree, migrations are supposed to modify database to new version on software update and there should be more than those two

Xertes0 replied:
Okay then what will we do about `/002_sample_data.sql`. This migration should only be applied on dev builds when we are testing the app. In production this migration should be omitted. In this case manually specifying the migrations would be better. We could do something like this:

```c++
#ifdef DEV_BUILD
migrator.addFile(i++, MAIN_DATABASE_MIGRATIONS_PATH "/002_sample_data.sql");
#endif
```

If we go with globing the entire directory, we need a way to filter all dev migrations.

Xertes0 replied:
> I disagree, migrations are supposed to modify database to new version on software update and there should be more than those two

Nothing stops us from adding a new migration on update, no matter which approach we choose.

papaj-na-wrotkach replied:
> Okay then what will we do about `/002_sample_data.sql`. This migration should only be applied on dev builds when we are testing the app. In production this migration should be omitted. In this case manually specifying the migrations would be better. We could do something like this:
> 
> ```c++
> #ifdef DEV_BUILD
> migrator.addFile(i++, MAIN_DATABASE_MIGRATIONS_PATH "/002_sample_data.sql");
> #endif
> ```
> 
> If we go with globing the entire directory, we need a way to filter all dev migrations.

This migration should be removed tbh and all the files from the migration directory should be applied in ascending order.

papaj-na-wrotkach replied:
We could also assume that all files in that directory are SQL files (eg. some may not have `.sql` extension if user adds it) and dispose of views... I just wanted to use it here but not using it should make the things easier to sort (no sort for views pipe syntax even in C++23 atm)

papaj-na-wrotkach replied:
resolved in recent changes

Xertes0 commented on server/src/dto/PageDto.hh:
```diff
@@ -0,0 +1,25 @@
+
+        DTO_FIELD(UInt64, count);
+
+        DTO_FIELD(Vector < T >, items);
```
Formatting bug? We use `Vector<T>` everywhere else.

papaj-na-wrotkach replied:
Yes, automatic formater issue when not using `oatpp::` prefix...

Xertes0 commented on server/src/dto/UserDto.hh:
```diff
@@ -0,0 +1,32 @@
+
+        DTO_FIELD(String, password, "password");
+
+        DTO_FIELD(Int32, role, "role"); // 0 - admin, 1 - user //TODO: enum
```
Should probably be done before merging.

papaj-na-wrotkach replied:
fixed in 19466743ef838762aa40d8d8a3bb9af33c8a3dfb

Xertes0 commented on server/src/main.cc:
```diff
@@ -1,45 +1,54 @@
+    );
+
+    auto const host = appComponent.serverConnectionProviderComponent.getObject()->getProperty("host").toString();
+    auto const address = "http://" + (host->find(":") == std::string::npos ? host : "[" + host + "]");
```
Could you make a comment on what is happening here?

papaj-na-wrotkach replied:
If it's not readable you should see the modified file only (without changes). I just edited the file to include all controllers and changed their names so git doesn't properly recognize the changes. I also removed the `::` prefix everywhere I could.

L37 and L38 generates the URL address of the server for the log. L38 adds the square braces if the host address is ipv6.

You may want to see [my comment above](https://github.com/mnurzyns/ParkSpaceSharing/pull/36/files/eaf878e5106238861b3c7c1f7a74ae917f92e110#r1179564371)

Xertes0 replied:
I'm talking about this line 38:
```
auto const address = "http://" + (host->find(":") == std::string::npos ? host : "[" + host + "]");
```
Why are you wrapping the host address in square brackets when it contains a colon?

Xertes0 replied:
> L38 adds the square braces if the host address is ipv6.

Yes, that is what I meant. I didn't know about the square-bracket-wrapping-for-ipv6 thingy.

Xertes0 replied:
Maybe we should add a comment for that? Just to mention it's for IPv6, because I don't thing everybody knows about it either.

Xertes0 commented on server/src/main.cc:
```diff
@@ -1,45 +1,54 @@
 
-    OATPP_COMPONENT(::std::shared_ptr<::oatpp::provider::Provider<::oatpp::sqlite::Connection>>, database_connection_provider);
-    database_connection_provider->stop();
+    OATPP_COMPONENT(std::shared_ptr<oatpp::provider::Provider<oatpp::sqlite::Connection>>, databaseConnectionProvider);
```
Shouldn't database provider be run before the server?

Xertes0 commented on CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@ cmake_minimum_required(VERSION ${CMAKE_REQUIRED_VERSION})
 
 option(CONAN "Use conan package manager" ON)
 
+#set(CMAKE_BUILD_TYPE Release)
```
This comment should be removed.

papaj-na-wrotkach replied:
Yes, but do you mean to remove the line or just uncomment it? I commented it out because it made my CLion use the `Release` build type even though I set it to `Debug`.

Xertes0 replied:
> Yes, but do you mean to remove the line or just uncomment it?

Remove the line.

> I commented it out because it made my CLion use the `Release` build type even though I set it to `Debug`.

Why not just change the build type in CLion?

papaj-na-wrotkach replied:
I changed it but the line overridden my CLion config

Xertes0 commented on README.md:
Maybe we should also note sqlite3 with fts5 support? Unless it's built in by default.

papaj-na-wrotkach replied:
> As of [version 3.9.0](https://www.sqlite.org/releaselog/3_9_0.html) (2015-10-14), FTS5 is included as part of the SQLite [amalgamation](https://www.sqlite.org/amalgamation.html).

Quoted from the [official SQLite page](https://www.sqlite.org/fts5.html#building_fts5_as_part_of_sqlite)

Xertes0 commented on server/src/service/AuthService.hh:
```diff
@@ -0,0 +1,34 @@
+#include "dto/SignInDto.hh"
+#include "dto/SignUpDto.hh"
+
+using Status = oatpp::web::protocol::http::Status;
```
Should be moved into the namespace.

Xertes0 commented on server/src/service/OfferService.cc:
```diff
@@ -0,0 +1,135 @@
+            OATPP_ASSERT_HTTP(false, Status::CODE_409, "Offer already exists")
+        }
+        catch (oatpp::web::protocol::http::HttpError &e) {
+            if (e.getInfo().status != Status::CODE_404) { throw; }
```
Again, why the if statement?

papaj-na-wrotkach replied:
see [this](https://github.com/mnurzyns/ParkSpaceSharing/pull/36/files/eaf878e5106238861b3c7c1f7a74ae917f92e110#r1179550193)

Xertes0 commented on server/src/service/OfferService.cc:
```diff
@@ -0,0 +1,135 @@
+            oatpp::UInt64 const &limit,
+            oatpp::UInt64 const &offset
+    ) {
+        std::string const queryTableFts = " FROM offer_fts";
+        std::string const queryFilters = query->empty() ? std::string{} : " WHERE offer_fts MATCH :query";
```
We can use `std::string_view` or `char const*` to avoid allocating heap memory.

papaj-na-wrotkach replied:
Can you show me what do you mean?

Xertes0 replied:
```c++
std::string_view queryTableFts = " FROM offer_fts";
std::string_view queryFilters = query->empty() ? "" : " WHERE offer_fts MATCH :query";
```

papaj-na-wrotkach replied:
fixed in 37a52bdc4aaaef38964d8bb8634d82a981c667a8

Xertes0 commented on server/src/service/OfferService.hh:
```diff
@@ -0,0 +1,59 @@
+        OATPP_COMPONENT(std::shared_ptr<database::MainDatabase>, database_);
+
+    public:
+        static std::shared_ptr<OfferService>
+        createShared() {
+            return std::make_shared<OfferService>();
+        }
```
Should be moved to a .cc file.

Xertes0 commented on server/src/service/OfferService.hh:
```diff
@@ -0,0 +1,59 @@
+                oatpp::Object<dto::OfferDto> const &dto
+        );
+
+        oatpp::Object<dto::OfferDto>
+        getOne(
+                oatpp::UInt64 const &id
+        );
+
+        oatpp::Object<dto::OfferPageDto>
+        search(
+                oatpp::String const &query,
+                oatpp::UInt64 const &limit = 20UL,
+                oatpp::UInt64 const &offset = 0UL
+        );
+
+        oatpp::Object<dto::OfferDto>
+        putOne(
+                oatpp::Object<dto::OfferDto> const &dto
+        );
+
+        oatpp::Object<dto::OfferDto>
+        patchOne(
+                oatpp::UInt64 const &id,
+                oatpp::Object<dto::OfferDto> const &dto
+        );
+
+        oatpp::Object<dto::StatusDto>
+        deleteOne(
+                oatpp::UInt64 const &id
+        );
```
Do not pass primitives as `const&`.

papaj-na-wrotkach replied:
but clang tidy...

papaj-na-wrotkach replied:
Also, is `oatpp::UInt64` a primitive? Is it only an Alias or sth more?

Xertes0 replied:
Okay never mind.
Turns out that UInt64 is an alias for a class that inherits from 2 other classes, and it holds a shared_ptr to the actual uin64.

Xertes0 replied:
`const&` should stay.

Xertes0 commented on server/src/service/OfferService.hh:
```diff
@@ -0,0 +1,59 @@
+        oatpp::Object<dto::OfferPageDto>
+        search(
+                oatpp::String const &query,
+                oatpp::UInt64 const &limit = 20UL,
+                oatpp::UInt64 const &offset = 0UL
```
`unsigned long` is not always 64 bits.
That is one of the reasons why we need CI and static analysis on Windows and MSVC.

papaj-na-wrotkach replied:
so what should we do?

Xertes0 replied:
We can either leave it as `oatpp::UInt64 const &offset = 0`, or use `std::uint64_t{0}`.
The first option is probably better.

The issue with leaving it as `20UL` is that if we try to compile it with MSVC on Windows, it will fail because of conversion error. (unsigned long on Windows is 32bit not 64bit)

papaj-na-wrotkach replied:
Can we leave it as `0U`?

papaj-na-wrotkach replied:
nvm, it won't compile either way:
```
error: conversion from ‘int’ to ‘const oatpp::data::mapping::type::UInt64’ {aka ‘const oatpp::data::mapping::type::Primitive<long unsigned int, oatpp::data::mapping::type::__class::UInt64>’} is ambiguous
   38 |            UInt64 const& offset = 0);
```

Xertes0 replied:
I tried some things, but:
```c++
UInt64 const& limit = std::uint64_t{20},
UInt64 const& offset = std::uint64_t{0});
```
is the only thing that worked.

Xertes0 replied:
Now the funny thing is that the only reason why `UInt64 const& offset = 0` doesn't work is, because `0` can be both `uint64_t` and `nullptr_t`. That's why it's ambiguous.

papaj-na-wrotkach replied:
I forgor to mention
https://github.com/oatpp/oatpp/issues/449

papaj-na-wrotkach replied:
I'll regex replace it for now

papaj-na-wrotkach replied:
fixed in 9f4496952fb5865b3c4dc90aeb534af7df5bf423

Xertes0 commented on server/src/service/PlaceService.cc:
This is more or less a copy of `OfferService.cc`. The same comments apply.

Xertes0 commented on server/src/service/PlaceService.hh:
This is more or less a copy of `OfferService.hh`. The same comments apply.

Xertes0 commented on server/src/service/UserService.cc:
This is more or less a copy of `OfferService.cc`. The same comments apply.

Xertes0 commented on server/src/service/UserService.hh:
This is more or less a copy of `OfferService.hh`. The same comments apply.

Xertes0 commented:
Before we merge we can cherry-pick #a3b2fcb for configuration support.

papaj-na-wrotkach commented:
Agreed, I wanted to do it anyways

papaj-na-wrotkach committed:
revert 1edec1116772ea429c4c55945d5d135e8e1dfa1f

papaj-na-wrotkach committed:
apply changes proposed by @Xertes0

papaj-na-wrotkach committed:
update README.md to reflect the changes

papaj-na-wrotkach committed:
use `[[nodiscard]]` in `AuthHandler`'s constructor

papaj-na-wrotkach commented on server/src/main.cc:
```diff
@@ -1,45 +1,54 @@
+    server::component::AppComponent appComponent;
 
-    OATPP_COMPONENT(::std::shared_ptr<::oatpp::web::server::HttpRouter>, router);
+    auto router = appComponent.httpRouterComponent.getObject();
```
I think it wasn't my change but @ArtMan0's. I don't know if it's a good practice to access a component by other component instead of calling the macro.

papaj-na-wrotkach replied:
I changed it back

papaj-na-wrotkach committed:
apply another batch of changes proposed by @Xertes0

Xertes0 committed:
Use `nlohmann_json` instead of `Boost`

Using Boost.JSON requires Boost >= 1.76 which is not available most
major distribution releases, and building it with conan would take a
lot of time. `nlohmann_json` makes a good alternative.

Because switching between json libraries requires changing only 1 line
of C++ code (`using Traits` in JWT.hh), we can add a CMake option to
let users compile our application with whatever supported
implementation they desire.

Currently supported JSON implementations by `jwt-cpp` are:
 - boost-json
 - danielaparker-jsoncons
 - kazuho-picojson
 - nlohmann-json

papaj-na-wrotkach commented on server/src/auth/JWT.cc:
```diff
@@ -22,8 +22,8 @@ namespace server::auth {
-                .set_payload_claim("userId", boost::json::value(payload->userId))
-                .set_payload_claim("role", boost::json::value(payload->role))
+                .set_payload_claim("userId", Traits::value_type{*payload->userId})
+                .set_payload_claim("role", Traits::value_type{*payload->role})
```
You shouldn't do that
use equivalent of `nlohmann_json` - sth like `nlohmann_json::json::value`

Xertes0 replied:
That is what `Traits::value_type` aliases to.

papaj-na-wrotkach replied:
Okay? But the alias is to the `jwt::traits::nlohmann_json::value_type` not to `nlohmann::json::value`...

Xertes0 replied:
> Okay? But the alias is to the `jwt::traits::nlohmann_json::value_type` not to `nlohmann::json::value`...

And what is the issue with that?

papaj-na-wrotkach replied:
nvm, they used `.set_payload_claim("sample", jwt::claim(std::string("test")))`
we should also use `jwt::claim()` like this:
```c++
.set_payload_claim("userId", jwt::claim{*payload->userId})
.set_payload_claim("role", jwt::claim{*payload->role})
```

Xertes0 replied:
Okay, will you change this in the next commit?

papaj-na-wrotkach replied:
yes

papaj-na-wrotkach committed:
use `jwt::claim` instead of `Traits::value_type`

papaj-na-wrotkach committed:
get rid of `std::ranges`, use C++20

papaj-na-wrotkach committed:
sort migrations by filename

Xertes0 committed:
Configuration file support

Adds 'server::config' struct to hold application's configuration read
from 'config.toml' file.

Currently the application only looks for the file in the current working
directory. If the file does not exist, it will be created with the
default values.

Xertes0 committed:
Merge branch `main` into dev/refactor

Xertes0 committed:
[Actions] Prevent useless 'Check for changes' runs

Xertes0 committed:
CMake install `server` target

`install` target also installs the oatpp swagger resources into
`/usr/local/share/<PROJECT_NAME>/`.

Xertes0 committed:
Oatpp swagger resource path as config variable

If the variable in config points to an invalid directory, the
application will try to use the directory from `oatpp-swagger`
conan package.

Xertes0 committed:
[Docker] cache build output for faster rebuilds

Caches CMake and Conan build output.

Makes use of CMake `install` target.

Xertes0 committed:
server::config swagger_res_path not loading fix

papaj-na-wrotkach commented on .gitignore:
```diff
@@ -36,4 +36,5 @@ build/*
 
-CMakeCache.txt
+**/config.toml
+/main.sqlite3
```
`main.sqlite3` should be `main.db`

Xertes0 replied:
I changed it to `main.sqlite3`.

papaj-na-wrotkach replied:
Please change it back to `main.db` - it's simpler and shorter

Xertes0 replied:
Isn't `.sqlite3` a bit more descriptive?

papaj-na-wrotkach replied:
imo too long

papaj-na-wrotkach replied:
if you want to compromise - some ppl use `.db3` but I don't think that's a good idea

Xertes0 replied:
Okay, I will change it to `.db` then.

papaj-na-wrotkach commented on server/src/component/AppComponent.hh:
```diff
@@ -42,8 +44,9 @@ namespace server::component {
+            auto const& config = server::config::get_instance();
             return oatpp::network::tcp::server::ConnectionProvider::createShared(
-                    {"::", 8000}, true); //TODO: read from config
+                {config.bind, config.port}, true);
```
Add fallback values?

papaj-na-wrotkach commented on server/src/component/AppComponent.hh:
```diff
@@ -23,7 +24,8 @@ namespace server::component {
          * */
         OATPP_CREATE_COMPONENT(std::shared_ptr<auth::JWT>, jwtComponent)([] {
-            return std::make_shared<auth::JWT>("secret", "issuer"); //TODO: read from config, add expiration time
+            auto const& config = server::config::get_instance();
```
also fallback values?

Xertes0 committed:
Consisten logging usign OATPP_LOG

Also removed `::` in `config.cc` to comply with code style.

Xertes0 committed:
Fix server crashing on toml::find_or

papaj-na-wrotkach committed:
eeee... uhm... changeees! (a lot so be careful when merging/rebasing)

Xertes0 commented on server/src/JWT.hh:
```diff
@@ -0,0 +1,44 @@
+
+namespace server {
+
+using oatpp::Void, oatpp::Any, oatpp::String, oatpp::Int8, oatpp::UInt8,
+  oatpp::Int16, oatpp::UInt16, oatpp::Int32, oatpp::UInt32, oatpp::Int64,
+  oatpp::UInt64, oatpp::Float32, oatpp::Float64, oatpp::Boolean, oatpp::Object,
+  oatpp::Enum, oatpp::Vector, oatpp::List, oatpp::UnorderedSet,
+  oatpp::UnorderedMap;
```
If you want to use `using` then please replace it with:
```c++
using namespace oatpp::data::mapping::type;
using namespace server::dto;
```

papaj-na-wrotkach replied:
clang tidy disallows using namespaces as you probably never use all members of the namespace

Xertes0 replied:
> clang tidy disallows using namespaces

Some clang-tidy checks shouldn't be taken as a 'rule' but more of a 'recommendation'. Al least that's how I view them.

> as you probably never use all members of the namespace

We already don't use most of the things that are included here so it won't make much difference. Mind that I don't want to include all of `oatpp` namespace, but only the type aliases.

papaj-na-wrotkach replied:
lmao
```
error: ‘dto’ is not a namespace-name
   12 | using namespace server::dto;
```

papaj-na-wrotkach replied:
And this only happens here, in `JWT.hh` - not in other places (eg `PlaceController`)

papaj-na-wrotkach replied:
ah, bc we don't use any dto here

Xertes0 replied:
Oh, we don't need `server::dto` namespace in this file. I did include it because I saw you did similar thing with all the dtos in a different file.

Xertes0 replied:
Yes, the dtos aren't included so `server::dto` is not a 'namespace-name', because this name doesn't exist in this file at all.

Xertes0 commented on server/src/controller/PlaceController.hh:
```diff
@@ -0,0 +1,249 @@
+                          "Required parameter not provided")
+
+        try {
+            OATPP_ASSERT_HTTP(
+              auth_object->role == 0 ||
+                (((dto->owner_id == nullptr &&
+                   (dto->owner_id = auth_object->user_id)) && // NOLINT
+                  auth_object->user_id == service_.getOne(dto->id)->owner_id) ||
+                 dto->owner_id == auth_object->user_id),
+              Status::CODE_403,
+              "Cannot create or modify place of another user as a regular user")
```
This is unreadable. Also the bug from previous conversation still exists.

Xertes0 replied:
Isn't it possible for a user to PUT a dto with NULL owner_id? Or does oatpp require that all of the dto's fields are non-null.

papaj-na-wrotkach replied:
it shouldn't contain that bug as if the `dto->owner_id == nullptr` then `dto->owner_id = auth_object->user_id` (notice single `=`) and `auth_object->user_id` is not null bc of `AUTHORIZATION` macro

papaj-na-wrotkach replied:
Oh maybe you meant to assert `dto->owner_id == auth_object->user_id` before using that `getOne` method? Then it would make sense

Xertes0 replied:
> it shouldn't contain that bug as if the `dto->owner_id == nullptr` then `dto->owner_id = auth_object->user_id` (notice single `=`) and `auth_object->user_id` is not null bc of `AUTHORIZATION` macro

But if an administrator makes that request `auth_object->role == 0`, then the later `dto->owner_id = auth_object->user_id` is not evaluated. And afaik `dto->owner_id` can be NULL.

As to the previous bug. It existed because `dto->owner_id == auth_object->user_id` was evaluated after `service_.getOne()`, and that is still the case.

Xertes0 replied:
Okay, I didn't notice the assert few lines above. So I see we *do* make sure that all of the dto fields are non-NULL. That means that `(dto->owner_id == nullptr && (dto->owner_id = auth_object->user_id))` is redundant, because we know that `dto->owner_id` has to contain a value.

And because here `dto->owner_id == nullptr` is always false, `service_.getOne()` will never be called, so the bug does not exist anymore.

Doesn't this thing simplify to just `auth_object->role == 0 || dto->owner_id == auth_object->user_id`?

papaj-na-wrotkach replied:
> `dto->owner_id` has to contain a value

that's not true, checks above only check if all fields which are marked as `NOT NULL` in db are not null. `owner_id` is a special case which we check in the second macro

Xertes0 replied:
> > `dto->owner_id` has to contain a value
> 
> that's not true, checks above only check if all fields which are marked as `NOT NULL` in db are not null. `owner_id` is a special case which we check in the second macro

To clarify I was talking about the asserts on the begging of the function:
```c++
OATPP_ASSERT_HTTP(dto->id && dto->owner_id && dto->address && dto->latitude && dto->longitude, Status::CODE_400, "Required parameter not provided")
```

So `dto->owner_id` *has* to contain a value.
And thus the whole:
```c++
((dto->owner_id == nullptr && (dto->owner_id = auth_object->user_id)) && auth_object->user_id == service_.getOne(dto->id)->owner_id)
```
is rendered redundant, because `dto->owner_id == nullptr` can't evaluate to `true`.

Xertes0 replied:
Maybe `dto->owner_id` wasn't supposed to be included in that assert?

papaj-na-wrotkach replied:
You're right, I must have poor eyesight because I overlooked that even though I was looking for it

papaj-na-wrotkach replied:
On my second thought I think, that we should not assign anything to DTO in controllers. I'll remake this

papaj-na-wrotkach replied:
And also `dto->id` can be null bc of `PRIMARY KEY` autoincrement in db

papaj-na-wrotkach replied:
Okay, nvm, it can be null in `createOne()` method, but it shouldn't be so in `replaceOne()`

papaj-na-wrotkach replied:
solved in a66ebab2994dbf1c60d2e6cd9538fa76c7e6b23e

Xertes0 commented on server/src/Config.cc:
```diff
@@ -0,0 +1,93 @@
+                   "The new config file will be placed in \"%s\".\n",
+                   config_path.c_str())
+        // This will construct the config with default values.
+        // See Config.hh fromToml()
```
It's `from_toml()`

papaj-na-wrotkach replied:
Yeah, renaming in my IDE broke it as I changed it back afterwards

papaj-na-wrotkach replied:
fixed

Xertes0 commented on server/src/Config.hh:
```diff
@@ -0,0 +1,97 @@
+    std::string bind;
+    std::uint16_t port;
+
+    std::string databasePath;
+
+    std::string jwt_secret;
+    std::string jwt_issuer;
+    std::uint64_t jwt_expire_after;
+
+    std::string swagger_res_path;
```
Inconsistent casing.

papaj-na-wrotkach replied:
fixed

Xertes0 commented on server/src/Config.hh:
```diff
@@ -0,0 +1,97 @@
+            .jwt_secret = toml::find_or<std::string>(jwt, "secret", "secret"),
+            .jwt_issuer = toml::find_or<std::string>(jwt, "issuer", "issuer"),
+            .jwt_expire_after = toml::find_or<std::uint64_t>(
+              jwt, "expire_after", 2592000), // 30 days
```
Isn't leaving `30*24*60*60` better? If you wanted to change it from 30 days to let's say 15, now you need to open up a calculator. Also it is not clear if this value is in seconds or minutes.

papaj-na-wrotkach replied:
use `system::chrono` in that case `¯\_(ツ)_/¯`

edit:
I did it

papaj-na-wrotkach committed:
change improper casing in `Config.hh`

papaj-na-wrotkach committed:
use `std::chrono` in `Config.hh`

papaj-na-wrotkach committed:
fix incorrect method name in comment

papaj-na-wrotkach committed:
use `OATPP_COMPONENT` to get `HttpRouter` instead of accessing it from `AppComponent`, more using to save space in code

Xertes0 committed:
Use `.db` for sqlite3 database extension

Agreed on in GitHub conversation.

papaj-na-wrotkach committed:
fix code style

Xertes0 committed:
JWT::JWT() pass string by value

papaj-na-wrotkach committed:
more code style fixes

papaj-na-wrotkach committed:
simplify the security assertions

papaj-na-wrotkach committed:
simplify usings (use `using namespace`

papaj-na-wrotkach committed:
pass `std::shared_ptr` by value instead of reference

papaj-na-wrotkach committed:
move `Payload` struct to a new file

papaj-na-wrotkach committed:
fix for unsigned long not always being 64-bit

papaj-na-wrotkach committed:
avoid allocating heap memory by reducing number of `std::string` objects

papaj-na-wrotkach committed:
make user role an enum

Xertes0 committed:
[Actions] Get rid of static analysis.

Last run took > 30min, and looking at the output, I just think it's
not worth it.

Xertes0 commented:
I removed static-analysis action in the latest commit.

> Last run took > 30min, and looking at the output, I just think it's
not worth it.

papaj-na-wrotkach committed:
fix namespace ending comment

Xertes0 committed:
[Actions] Do not install Qt

We probably won't use it in this project anymore.

Also removes some left over comment in `CMakeLists.txt`.

Xertes0 committed:
Update README.md

Remove unneeded qt dependencies from build prerequisites and replace
`Boost.JSON` with `nlohmann_json` in required dependencies.

Xertes0 committed:
CMakeLists.txt: redundant find_package() call
